eclair-core/src/main/scala/fr/acinq/eclair/balance/CheckBalance.scala

```diff
@@ -0,0 +1,295 @@
+package fr.acinq.eclair.balance
+
+import com.softwaremill.quicklens._
+import fr.acinq.bitcoin.{Btc, ByteVector32, SatoshiLong}
+import fr.acinq.eclair.MutualCloseStatus
+import fr.acinq.eclair.blockchain.bitcoind.rpc.ExtendedBitcoinClient
+import fr.acinq.eclair.channel.Helpers.Closing
+import fr.acinq.eclair.channel.Helpers.Closing.{CurrentRemoteClose, LocalClose, NextRemoteClose, RemoteClose}
+import fr.acinq.eclair.channel._
+import fr.acinq.eclair.transactions.DirectedHtlc.{incoming, outgoing}
+import fr.acinq.eclair.transactions.Transactions
+import fr.acinq.eclair.transactions.Transactions.{ClaimHtlcSuccessTx, ClaimHtlcTimeoutTx, HtlcSuccessTx, HtlcTimeoutTx}
+import fr.acinq.eclair.wire.protocol.UpdateFulfillHtlc
+
+import scala.concurrent.{ExecutionContext, Future}
+
+object CheckBalance {
+
+  /**
+   * For more fine-grained analysis, we count the in-flight amounts separately from the main amounts.
+   *
+   * The base assumption regarding htlcs is that they will all timeout. That means that we ignore incoming htlcs, and we
+   * count outgoing htlcs in our balance.
+   */
+  case class MainAndHtlcBalance(toLocal: Btc = 0.sat, htlcOut: Btc = 0.sat) {
+    val total: Btc = toLocal + htlcOut
+  }
+
+  /**
+   * In the closing staten some transactions may be published or even confirmed. They will be taken into account if we
+   * do a `bitcoin-cli getbalance` and we don't want to count them twice.
+   *
+   * That's why we keep track of the id of each transaction that pays us any amount. It allows us to double check from
+   * bitcoin core and remove any published transaction.
+   */
+  case class PossiblyPublishedBalance(toLocal: Map[ByteVector32, Btc] = Map.empty, htlcs: Map[ByteVector32, Btc] = Map.empty, htlcsUnpublished: Btc = 0.sat) {
+    val totalToLocal: Btc = toLocal.values.map(_.toSatoshi).sum
+    val totalHtlcs: Btc = htlcs.values.map(_.toSatoshi).sum
+    val total: Btc = totalToLocal + totalHtlcs + htlcsUnpublished
+  }
+
+  /**
+   * Mutual close transactions are always immediately published, and will be taken into account by bitcoin core, so we
+   * don't count them in the balance.
+   */
+  case class ClosingBalance(
+                             localCloseBalance: PossiblyPublishedBalance = PossiblyPublishedBalance(),
+                             remoteCloseBalance: PossiblyPublishedBalance = PossiblyPublishedBalance(),
+                             unknownCloseBalance: MainAndHtlcBalance = MainAndHtlcBalance()
+                           ) {
```

Comment:
nit: I really find the indentation weird (same for other case classes below), I don't get why the formatter does that...

Reply:
The formatter is innocent, I bear the sole responsibility I'm afraid. Would you prefer:

```scala
case class ClosingBalance(localCloseBalance: PossiblyPublishedBalance = PossiblyPublishedBalance(),
                          remoteCloseBalance: PossiblyPublishedBalance = PossiblyPublishedBalance(),
                          unknownCloseBalance: MainAndHtlcBalance = MainAndHtlcBalance()) {
```

Reply:
Yes that looks better!

Reply:
Done with 4b78392.